### PR TITLE
Improve Docker dependencies code

### LINKED ISSUE #386: strange behaviour with .dockerignore
microservices example
### Expected behavior
should ignore files stated in `.dockerignore` without `**`, like `app.go`
### Actual behavior
only ignores files which have `**`, like `**/app.go`
### Information
Given microservices example
And `.dockerignore` file in same folder as `Dockerfile` with content `app.go`
When `skaffold dev` is run 
It does not ignore `app.go`
```
Skaffold version: 0.4.0
Operating system: macOS
Content of skaffold.yaml: 
apiVersion: skaffold/v1alpha2
kind: Config
build:
  artifacts:
  - imageName: gcr.io/k8s-skaffold/leeroy-web
    workspace: ./leeroy-web/
    docker: {}
  - imageName: gcr.io/k8s-skaffold/leeroy-app
    workspace: ./leeroy-app/
    docker: {}
  local: {}
deploy:
  kubectl:
    manifests:
    - ./leeroy-web/kubernetes/*
    - ./leeroy-app/kubernetes/*
```

### Steps to reproduce the behavior

  1. echo 'app.go' > leeroy-app/.dockerignore 
  2. skaffold dev should fail

### LINKED ISSUE #462: .dockerignore isn't watched for changes
### Expected behavior
Changes to `.dockerignore` should be reflected immediately.

### Actual behavior
Need to restart Skaffold to pickup changes in `.dockerignore`.

### Information

```
Skaffold version: v0.4.0
Operating system: 10.13.3
Content of skaffold.yaml:
apiVersion: skaffold/v1alpha2
kind: Config
build:
  artifacts:
  - imageName: gcr.io/wdenniss-gke-test17/nodedemo
deploy:
  kubectl:
    manifests:
      - k8s-*
profiles:
  - name: gcb
    build:
      googleCloudBuild:
        projectId: wdenniss-gke-test17
```

### Steps to reproduce the behavior

  1. Add a file to `.dockerfile`
  2. Modify that file
  3. Observe that Skaffold still triggers despite the file being ignored

### PR DESCRIPTION
**Pull request by dgageot**
Fixes #386
Also Fixes #462

### CODE REVIEW

File: pkg/skaffold/docker/context_test.go [hunk 1]
```diff
@@ -19,51 +19,59 @@ package docker
...
 )
 
 func TestDockerContext(t *testing.T) {
+	tmpDir, cleanup := testutil.TempDir(t)
+	defer cleanup()
+
+	os.Mkdir(filepath.Join(tmpDir, "files"), 0750)
+	ioutil.WriteFile(filepath.Join(tmpDir, "files", "ignored.txt"), []byte(""), 0644)
+	ioutil.WriteFile(filepath.Join(tmpDir, "files", "included.txt"), []byte(""), 0644)
+	ioutil.WriteFile(filepath.Join(tmpDir, ".dockerignore"), []byte("**/ignored.txt"), 0644)
```
Comment (tejal29) [inline pkg/skaffold/docker/context_test.go]: Can we add a test case for #386. 
Add a line to .dockerignore with "alsoignored.txt" and create a file alsoignored.txt.

File: pkg/skaffold/docker/parse_test.go [hunk 2]
```diff
@@ -159,7 +159,7 @@ func TestGetDockerfileDependencies(t *testing.T) {
 			description: "add dependency",
 			dockerfile:  addDockerfile,
 			workspace:   "docker",
-			expected:    []string{"docker/Dockerfile", "docker/nginx.conf"},
+			expected:    []string{"Dockerfile", "nginx.conf"},
```
Comment (tejal29) [inline pkg/skaffold/docker/parse_test.go]: This is worry some for me. I don't have a lot of context, but earlier GetDockerfileDependencies used to return absolute paths.
Now you dont.  Would this not break the workflow.

Comment (dgageot) [reply pkg/skaffold/docker/parse_test.go]: It shouldn't but yes, that's a big change. One that's needed

File: pkg/skaffold/docker/context_test.go (hunk 1 above)
Comment (dgageot) [reply pkg/skaffold/docker/context_test.go]: I'll do that, thanks for the idea!

File: pkg/skaffold/docker/parse_test.go (hunk 2 above)
Comment (dgageot) [reply pkg/skaffold/docker/parse_test.go]: We need to keep things as relative as possible. Each time we switch to absolute paths, it's a mess

**Comment by dgageot** [comment]
@tejal29 I've added the test. Regarding the paths, I spend quite some time trying to fix our path madness and keeping relative paths as much as possible was the best solution I found. There's more work to be done on the kubectl deployer side.

File: pkg/skaffold/docker/parse_test.go (hunk 2 above)
Comment (tejal29) [reply pkg/skaffold/docker/parse_test.go]: yes i agree absolute paths are a mess.
Will take one more look.

**Pull request merged**